deckhouse-controller/pkg/registry/deckhouse.go

```diff
@@ -0,0 +1,230 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package registry
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/ettle/strcase"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	regTransport "github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"gopkg.in/yaml.v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/utils"
+	"github.com/deckhouse/deckhouse/go_lib/dependency"
+	"github.com/deckhouse/deckhouse/go_lib/dependency/cr"
+	"github.com/deckhouse/deckhouse/go_lib/libapi"
+)
+
+type DeckhouseService struct {
+	dc dependency.Container
+
+	registry        string
+	registryOptions []cr.Option
+}
+
+func NewDeckhouseService(registryAddress string, registryConfig *utils.RegistryConfig) *DeckhouseService {
+	return &DeckhouseService{
+		dc:              dependency.NewDependencyContainer(),
+		registry:        registryAddress,
+		registryOptions: utils.GenerateRegistryOptions(registryConfig),
+	}
+}
+
+func (svc *DeckhouseService) GetDeckhouseRelease(ctx context.Context, releaseChannel string) (*ReleaseMetadata, error) {
+	regCli, err := svc.dc.GetRegistryClient(path.Join(svc.registry, "release-channel"), svc.registryOptions...)
+	if err != nil {
+		return nil, fmt.Errorf("get registry client: %w", err)
+	}
+
+	img, err := regCli.Image(ctx, strcase.ToKebab(releaseChannel))
+	if err != nil {
+		if strings.Contains(err.Error(), string(regTransport.ManifestUnknownErrorCode)) {
+			err = errors.Join(err, ErrChannelIsNotFound)
+		}
+
+		return nil, fmt.Errorf("fetch image: %w", err)
+	}
+
+	releaseMetadata, err := svc.fetchReleaseMetadata(img)
+	if err != nil {
+		return nil, fmt.Errorf("fetch release metadata: %w", err)
+	}
+
+	if releaseMetadata.Version == nil {
+		return nil, fmt.Errorf("release metadata malformed: no version found")
+	}
+
+	return releaseMetadata, nil
+}
+
+func (svc *DeckhouseService) ListDeckhouseReleases(ctx context.Context) ([]string, error) {
+	regCli, err := svc.dc.GetRegistryClient(svc.registry, svc.registryOptions...)
+	if err != nil {
+		return nil, fmt.Errorf("get registry client: %w", err)
+	}
+
+	ls, err := regCli.ListTags(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("list tags: %w", err)
+	}
+
+	return ls, nil
+}
+
+type canarySettings struct {
+	Enabled  bool            `json:"enabled"`
+	Waves    uint            `json:"waves"`
+	Interval libapi.Duration `json:"interval"` // in minutes
+}
+
+type ReleaseMetadata struct {
```

Comment:
I guess it would be better to make public and use [this struct](https://github.com/deckhouse/deckhouse/blob/01ff165b7d38b483a2b9e39aa85dde9281741463/deckhouse-controller/pkg/controller/deckhouse-release/check_release.go#L638). Otherwise, sometime we will lose some new attribute

Reply:
done